File: tests/search.test.ts
import { describe, expect, test } from 'bun:test';
import { nameMatchIndices, searchCommands } from '../src/search';
import type { CommandEntry } from '../src/library';
import type { State } from '../src/state';

// Spec §3.1: match over name + description + command, name weighted highest,
// then description, then command. Empty query: MRU first (State keyed by id),
// never-used follow in file (array) order.

const commands: CommandEntry[] = [
  { id: 'c1', name: 'deploy prod', command: 'ssh prod-1 deploy.sh', description: 'Roll out to production' },
  { id: 'c2', name: 'tail logs', command: 'aws logs tail /ecs/api --follow', description: 'Tail ECS logs' },
  { id: 'c3', name: 'docker nuke', command: 'docker system prune -af', description: 'Remove unused containers' },
  { id: 'c4', name: 'list ports', command: 'lsof -iTCP -sTCP:LISTEN', description: 'Show listening processes' },
];

const names = (result: CommandEntry[]) => result.map((c) => c.name);

describe('empty query', () => {
  test('MRU first, never-used in file order after', () => {
    const state: State = {
      c3: { lastUsedAt: '2026-07-20T00:00:00Z' },
      c2: { lastUsedAt: '2026-07-23T00:00:00Z' },
    };
    expect(names(searchCommands(commands, state, ''))).toEqual([
      'tail logs',
      'docker nuke',
      'deploy prod',
      'list ports',
    ]);
  });

  test('no state at all keeps pure file order', () => {
    expect(names(searchCommands(commands, {}, ''))).toEqual([
      'deploy prod',
      'tail logs',
      'docker nuke',
      'list ports',
    ]);
  });
});

describe('fuzzy matching', () => {
  test('subsequence matches on the name', () => {
    expect(names(searchCommands(commands, {}, 'dpl'))).toContain('deploy prod');
  });

  test('non-matching commands are filtered out', () => {
    expect(searchCommands(commands, {}, 'zzzz')).toEqual([]);
  });

  test('a name hit outranks a description hit', () => {
    // "tail" is the name of one command and in the description of another
    const result = searchCommands(commands, {}, 'tail');
    expect(result[0]!.name).toBe('tail logs');
  });

  test('a description hit outranks a command-text hit', () => {
    // "listen" appears in b's description and a matches only in its command
    const cmds: CommandEntry[] = [
      { id: 'a', name: 'a', command: 'echo listening' },
      { id: 'b', name: 'b', command: 'echo x', description: 'listening things' },
    ];
    expect(searchCommands(cmds, {}, 'listening')[0]!.name).toBe('b');
  });
});

describe('nameMatchIndices', () => {
  test('returns the greedy subsequence positions in the name', () => {
    expect(nameMatchIndices('dpl', 'deploy prod')).toEqual(new Set([0, 2, 3]));
  });

  test('is case-insensitive', () => {
    expect(nameMatchIndices('DP', 'deploy prod')).toEqual(new Set([0, 2]));
  });

  test('null when the name does not match (row may have hit via description)', () => {
    expect(nameMatchIndices('zzz', 'deploy prod')).toBeNull();
  });

  test('null on an empty or whitespace query', () => {
    expect(nameMatchIndices('', 'deploy prod')).toBeNull();
    expect(nameMatchIndices('  ', 'deploy prod')).toBeNull();
  });
});

File: internal/search/search_test.go
package search

import (
	"testing"

	"github.com/luojiahai/potato/internal/library"
	"github.com/luojiahai/potato/internal/state"
)

// Spec §3.1: match over name + description + command, name weighted highest,
// then description, then command. Empty query: MRU first (State keyed by id),
// never-used follow in file (array) order.

func describe(s string) *string { return &s }

var commands = []library.Entry{
	{ID: "c1", Name: "deploy prod", Command: "ssh prod-1 deploy.sh", Description: describe("Roll out to production")},
	{ID: "c2", Name: "tail logs", Command: "aws logs tail /ecs/api --follow", Description: describe("Tail ECS logs")},
	{ID: "c3", Name: "docker nuke", Command: "docker system prune -af", Description: describe("Remove unused containers")},
	{ID: "c4", Name: "list ports", Command: "lsof -iTCP -sTCP:LISTEN", Description: describe("Show listening processes")},
}

func names(entries []library.Entry) []string {
	out := make([]string, 0, len(entries))
	for _, entry := range entries {
		out = append(out, entry.Name)
	}
	return out
}

func equal(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func TestEmptyQueryMRUFirst(t *testing.T) {
	s := state.State{
		"c3": {LastUsedAt: "2026-07-20T00:00:00Z"},
		"c2": {LastUsedAt: "2026-07-23T00:00:00Z"},
	}
	want := []string{"tail logs", "docker nuke", "deploy prod", "list ports"}
	if got := names(Commands(commands, s, "")); !equal(got, want) {
		t.Errorf("got %v, want %v", got, want)
	}
}

func TestEmptyQueryNoStateKeepsFileOrder(t *testing.T) {
	want := []string{"deploy prod", "tail logs", "docker nuke", "list ports"}
	if got := names(Commands(commands, state.State{}, "")); !equal(got, want) {
		t.Errorf("got %v, want %v", got, want)
	}
}

func TestSubsequenceMatchOnName(t *testing.T) {
	got := names(Commands(commands, state.State{}, "dpl"))
	found := false
	for _, name := range got {
		if name == "deploy prod" {
			found = true
		}
	}
	if !found {
		t.Errorf("got %v, want it to contain 'deploy prod'", got)
	}
}

func TestNonMatchingFilteredOut(t *testing.T) {
	if got := Commands(commands, state.State{}, "zzzz"); len(got) != 0 {
		t.Errorf("got %v, want none", names(got))
	}
}

func TestNameHitOutranksDescriptionHit(t *testing.T) {
	got := Commands(commands, state.State{}, "tail")
	if got[0].Name != "tail logs" {
		t.Errorf("got %q first, want 'tail logs'", got[0].Name)
	}
}

func TestDescriptionHitOutranksCommandHit(t *testing.T) {
	entries := []library.Entry{
		{ID: "a", Name: "a", Command: "echo listening"},
		{ID: "b", Name: "b", Command: "echo x", Description: describe("listening things")},
	}
	got := Commands(entries, state.State{}, "listening")
	if got[0].Name != "b" {
		t.Errorf("got %q first, want 'b'", got[0].Name)
	}
}

func TestNameMatchIndices(t *testing.T) {
	indices, ok := NameMatchIndices("dpl", "deploy prod")
	if !ok {
		t.Fatal("expected a match")
	}
	for _, i := range []int{0, 2, 3} {
		if !indices[i] {
			t.Errorf("index %d not matched: %v", i, indices)
		}
	}
	if len(indices) != 3 {
		t.Errorf("got %v, want exactly {0,2,3}", indices)
	}
}

func TestNameMatchIndicesIsCaseInsensitive(t *testing.T) {
	indices, ok := NameMatchIndices("DP", "deploy prod")
	if !ok || !indices[0] || !indices[2] || len(indices) != 2 {
		t.Errorf("got %v ok=%v, want {0,2}", indices, ok)
	}
}

func TestNameMatchIndicesMisses(t *testing.T) {
	for _, query := range []string{"zzz", "", "  "} {
		if _, ok := NameMatchIndices(query, "deploy prod"); ok {
			t.Errorf("query %q reported a match", query)
		}
	}
}
